unittests/AST/TensorFlow.cpp

```diff
@@ -99,3 +104,78 @@ TEST_F(TypeContainsTensorFlowValueTest, WorksForRecursiveTypes) {
   recursiveDecl->addMember(varDecl);
   EXPECT_FALSE(containsTensorFlowValue(recursiveType));
 }
+
+TEST_F(TypeContainsTensorFlowValueTest, TensorFlowType) {
+  EXPECT_TRUE(tf::isTensorHandle(tensorHandleType));
+  EXPECT_FALSE(tf::isTensorHandle(floatType));
+}
+
+TEST_F(TypeContainsTensorFlowValueTest, TensorHandleElementType) {
+  EXPECT_TRUE(tf::getTensorHandleElementType(floatType).isNull());
+  EXPECT_FALSE(tf::getTensorHandleElementType(tensorHandleType).isNull());
+}
+
+TEST_F(TypeContainsTensorFlowValueTest, ClassifyTensorFlowTypes) {
+  EXPECT_EQ(tf::TFValueKind::TensorHandle,
+            tf::classifyTensorFlowValue(tensorHandleType));
+
+  auto *resourceHandleDecl =
+      testContext.makeNominal<ClassDecl>("ResourceHandle");
+  Type resourceHandleType = resourceHandleDecl->getDeclaredInterfaceType();
+  EXPECT_EQ(tf::TFValueKind::ResourceHandle,
+            tf::classifyTensorFlowValue(resourceHandleType));
+
+  auto *variantHandleDecl = testContext.makeNominal<ClassDecl>("VariantHandle");
+  Type variantHandleType = variantHandleDecl->getDeclaredInterfaceType();
+  EXPECT_EQ(tf::TFValueKind::VariantHandle,
+            tf::classifyTensorFlowValue(variantHandleType));
+
+  EXPECT_EQ(tf::TFValueKind::Nope, tf::classifyTensorFlowValue(floatType));
+
+  auto *classDecl = testContext.makeNominal<ClassDecl>("SomeClass");
+  Type classType = classDecl->getDeclaredInterfaceType();
+  EXPECT_EQ(tf::TFValueKind::Nope, tf::classifyTensorFlowValue(classType));
+
+  // We test all functions defined around classifyTensorFlowValue
+  EXPECT_TRUE(tf::isTensorHandle(tensorHandleType));
+  EXPECT_FALSE(tf::isTensorHandle(floatType));
+
+  EXPECT_TRUE(tf::isOpaqueHandle(resourceHandleType) &&
+              tf::isOpaqueHandle(variantHandleType));
+  EXPECT_FALSE(tf::isOpaqueHandle(classType));
+
+  EXPECT_TRUE(tf::isTensorFlowValue(tensorHandleType) &&
+              tf::isTensorFlowValue(resourceHandleType) &&
+              tf::isTensorFlowValue(variantHandleType));
+  EXPECT_FALSE(tf::isTensorFlowValue(floatType));
+}
+
+TEST_F(TypeContainsTensorFlowValueTest, flattenTensorFlowValueAggregate) {
+  SmallVector<Type, 4> flattenedTensorFlowTypes;
+
+  auto *anyDecl = testContext.makeNominal<ClassDecl>("AnyHandle");
+  EXPECT_FALSE(tf::flattenTensorFlowValueAggregate(
+      anyDecl->getDeclaredInterfaceType(), flattenedTensorFlowTypes));
+  EXPECT_EQ(0ul, flattenedTensorFlowTypes.size());
```

Comment:
Could you check the contents of the returned types as well? I *think* you should be able to use matchers if you set up the actual and expected types to be canonical (See `Type->getCanonicalType`):
```
EXPECT_THAT(flattenedTensorFlowTypes, ::testing::ElementsAre(expected1, expected2, ...))
```

Reply:
Indeed it makes sense to check the contents of the vector, thank you for the pointers on how to do this.

Given that direct comparison is not possible on `Type` objects, I had to write my own `Matcher` to achieve this. So instead, I am doing something like:
```
EXPECT_THAT(flattenedTensorFlowTypes, ::testing::Each(CUSTOM_MATCHER))
```
Check out the next commit the details on my next commit pls :).